### DISCUSSION TITLE
No completion for asdf-vm commands

### DISCUSSION THREAD

**Discussion by Pilgrim1379** [marlonrichert/zsh-snap#131] [Q&A] [answered]
I've pretty much remodelled my zsh setup to mimick your personal zsh settings. I've even organised the structure of my files same as yours with the naming and all. Everything is working as expected except completion for the asdf-vm framework. 

I've installed asdf-vm via `znap source asdf-vm/asdf asdf.sh` and also have my fpath like so:
```
fpath=(
    $HOME/Git/asdf/completions(N)
    $HOMEBREW_PREFIX/share/zsh/site-functions
    $fpath[@]
)
```

However when I start `asdf ` and hit TAB, I don't get any completion for the command options. The completions file available at  `$HOME/Git/asdf/completions` as a file `_asdf`

**Comment by marlonrichert** [comment, marked as answer]
> also have my fpath like so:
> ```
> fpath=(
>     $HOME/Git/asdf/completions(N)
>     $HOMEBREW_PREFIX/share/zsh/site-functions
>     $fpath[@]
> )
> ```

That's no longer the recommended way to install completions. Remove that and run this once on the command line:
```
znap install asdf-vm/asdf
```
It will work after that.

**Reply by Pilgrim1379** [reply]
It worked thanks. These changes, are they documented anywhere. I'm sure I scoured through the docs but didn't see it mentioned. 

Also in my `zshrc` could I have done 
```
znap install asdf-vm/asdf
znap source asdf-vm/asdf
```
so I get the competions installed properly next time? I'm trying to create a set and forget script.

Not an issue but I noticed two ui's for listing completions. One lists the completions vertically with descriptions of what each does (which I prefer) and the other lists them horizontally with no descriptions at all.

So asdf-vm list's veritically with nice descriptions whereas pipx and pip for example only lists the completions horizontally. Is it somehing that I can control?

**Reply by marlonrichert** [reply]
> These changes, are they documented anywhere.

Yes, here: https://github.com/marlonrichert/zsh-snap/blob/main/README.md#install-executables-and-completion-functions

> ```zsh
> znap install asdf-vm/asdf
> ```

While you _can_ technically put that in your dotfiles, it will slow down your shell startup somewhat, since each time you run `znap install`, it scans the repo for new executables and completion functions. That's not a "free" operation. You're meant to just run it once on the command line.

> Not an issue but I noticed two ui's for listing completions. One lists the completions vertically with descriptions of what each does (which I prefer) and the other lists them horizontally with no descriptions at all.
> 
> So asdf-vm list's veritically with nice descriptions whereas pipx and pip for example only lists the completions horizontally. Is it somehing that I can control?

No, unfortunately not. This happens because `pipx` and `pip` use `bash`-style completions, which do not support descriptions. To get the vertical list with descriptions, you need a native `zsh` completion function, which `asdf` does provide, but `pipx` and `pip` do not.

**Reply by marlonrichert** [reply]
Oh, and thanks for the sponsorship! ❤️

**Reply by Pilgrim1379** [reply]
Got you. thanks a bunch!
